core/src/main/java/com/linecorp/armeria/common/stream/FuseableStreamMessage.java

```diff
@@ -295,8 +295,8 @@ public void onNext(Object item) {
                 if (result != null && item != result) {
                     StreamMessageUtil.closeOrAbort(result, ex);
                 }
-                upstream.cancel();
                 onError(ex);
+                upstream.cancel();
```

Comment:
So in general, do we always cancel the upstream before propagating the error downstream now?

Reply:
If we cancel the upstream first, `cancel()` may recursively call `downstream.onError(CancelledSubscriptionException)` which could prevent the error from being propagated to `downstream`.

So I think `onError(ex)` should be invoked first to propagate the cause to the downstream that will be eventually passed to `ServerErrorHandler` and `RequestLog.*Cause()`

Reply:
So this change is unrelated to the leak, am I understanding correctly?

Reply:
Right. I found the bug while fixing broken tests after removing `CollectingSubscriberAndSubscription` and related code.